src/patterns.jl

```diff
@@ -0,0 +1,131 @@
+"""
+    AbstractPattern
+
+Abstract supertype of all sparsity pattern representations.
+
+## Type hierarchy
+```
+AbstractPattern
+├── AbstractVectorPattern: used in GradientTracer, ConnectivityTracer
+│   └── IndexSetVectorPattern
```

Comment:
I think it would be a good idea to flesh out an array-backed pattern in this same PR, to anticipate potential problems like carrying the size (see below)

Reply:
I'd rather do this in another PR since it requires input enumeration.
I have some ideas for this I hinted at in #28. I'd also try to support sparse-matrix inputs, the need for which popped up in #131.